### PR TITLE
Maybe run autogen as part of freetype install

### PR DESCRIPTION
**Pull request by wqh17101**
run autogen.sh before configure freetype
https://github.com/matplotlib/matplotlib/issues/22754

## PR Summary
run autogen.sh before configure freetype
## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by QuLogic** [comment]
Why? We use a release tarball, so autotools is not necessary.

File: setupext.py
Comment (tacaswell) [inline setupext.py:630]: I think this logic should be "If the results of autogen.sh is not there, run autotools"

**Comment by tacaswell** [comment]
We do document that the user can find some mechanism to get the source in the right place, I think it is fair to support the "I got this directly from git" case natively (but agree that we should not depend on autotools in general!).

File: setupext.py
Comment (wqh17101) [reply setupext.py:630]: I don't think that logic is needed for  that i think  if autogen.sh is not there, which means that it is not necessary for the project.

**Comment by wqh17101** [comment]
> Why? We use a release tarball, so autotools is not necessary.

You can see my issue. Sometimes you must fix vulnerabilities by patch when it is not released now.

**Comment by wqh17101** [comment]
It seems that it appeared many error which is not caused my PR, so is this the right branch for me to merge to?

**Comment by tacaswell** [comment]
The azure windows failures are unrelated, but the rest of the failures are caused by this PR.

If we have a `configure` then we do not need to run `sh autogen.sh` which uses autotools internally.

**Review by tacaswell — CHANGES_REQUESTED** [review]
Only run the autogen script if we need to.

We can not pick up a build-time dependency on autotools.

**Comment by wqh17101** [comment]
> Only run the autogen script if we need to.
> 
> We can not pick up a build-time dependency on autotools.

You mean trying to build without `sh autogen.sh` first, and rerun `sh autogen.sh and build` when there was an exception?

**Comment by tacaswell** [comment]
> You mean trying to build without sh autogen.sh first,

Yes, or just check if `configure` exists?

**Comment by wqh17101** [comment]
> If we have a `configure` then we do not need to run `sh autogen.sh` which uses autotools internally.
You can see the freetype source code , `configure`  and `autogen.sh` exist at the same time.So i do not think this logic is right.
Also you can see  https://gitlab.freedesktop.org/freetype/freetype/-/blob/master/README.git#L36 and https://gitlab.freedesktop.org/freetype/freetype/-/blob/master/.gitlab-ci.yml#L94 , the right way to build freetype from source is running `autogen.sh` first.

**Comment by wqh17101** [comment]
So i think run `autogen.sh` will be both right for the source code and release tarball

**Comment by wqh17101** [comment]
Oh , i see. you mean that sometimes , the env may not contain the autotools.
So the right logic i think is that run autogen.sh first, and continue if exceptions appear.

File: setupext.py
Comment (tacaswell) [inline setupext.py:632]: In almost all cases this will be a spurious warning.  We have not been using autotools from at least 2015.

Can you please point to documentation from freetype that when installing from the release tarball it is expected that running `autogen.sh` is required?

Comment (wqh17101) [reply setupext.py]: Alright, i find that autogen.sh is used to generate `configure.ac`.So just run it if this file does not exist.

File: setupext.py
Comment (oscargus) [inline setupext.py:633]:
```suggestion
                print(f"{configure_ac} does not exist. Using sh autogen.sh to generate.")
```

Comment (jklymak) [reply setupext.py]: please let commenters resolve, thanks!

**Review by wqh17101 — COMMENTED** [review]
changed

**Review by tacaswell — APPROVED** [review]

**Comment by wqh17101** [comment]
Although i modified the code for the limitation of max_line_length of flake8, it is better to set `max_line_length=100 or 120` for now i think.

**Comment by oscargus** [comment]
Power cycling to restart the tests.

**Comment by tacaswell** [comment]
I want to leave final review of this to @QuLogic .

**Comment by jklymak** [comment]
ping @QuLogic this is waiting for your review.

**Review by QuLogic — APPROVED** [review]

**Pull request merged**
